### PR TITLE
Fix.2282.fix.OIDC.totp.flow

### LINKED ISSUE #2272: API Session Refresh Results in `native` audience JWT
On API session renewal on HA version v1.1.8, after 30m the tunneler reports:

```
[2024-07-31T15:51:46.919Z]   ERROR tunnel-cbs:ziti_tunnel_cbs.c:103 on_ziti_connect() ziti dial failed: connection is closed
[2024-07-31T15:51:47.460Z]   ERROR ziti-sdk:connect.c:964 connect_reply_cb() conn[4.15/cjRT9JFN/Connecting] failed to connect, reason=invalind audience, expected an instance of openziti, got [native]
[2024-07-31T15:51:47.460Z]   ERROR tunnel-cbs:ziti_tunnel_cbs.c:103 on_ziti_connect() ziti dial failed: connection is closed
[2024-07-31T15:51:48.007Z]   ERROR ziti-sdk:connect.c:964 connect_reply_cb() conn[4.16/eJNwcOi1/Connecting] failed to connect, reason=invalind audience, expected an instance of openziti, got [native]
[2024-07-31T15:51:48.007Z]   ERROR tunnel-cbs:ziti_tunnel_cbs.c:103 on_ziti_connect() ziti dial failed: connection is closed
[2024-07-31T15:51:48.550Z]   ERROR ziti-sdk:connect.c:964 connect_reply_cb() conn[4.17/CurzKOB_/Connecting] failed to connect, reason=invalind audience, expected an instance of openziti, got [native]
[2024-07-31T15:51:48.550Z]   ERROR tunnel-cbs:ziti_tunnel_cbs.c:103 on_ziti_connect() ziti dial failed: connection is closed
[2024-07-31T15:51:49.092Z]   ERROR ziti-sdk:connect.c:964 connect_reply_cb() conn[4.18/reWy12Jh/Connecting] failed to connect, reason=invalind audience, expected an instance of openziti, got [native]
[2024-07-31T15:51:49.093Z]   ERROR tunnel-cbs:ziti_tunnel_cbs.c:103 on_ziti_connect() ziti dial failed: connection is closed
```

Where the audience of `native` is not expected.

### PR DESCRIPTION
**Pull request by andrewpmartinez**
This also fixes #2272

### CODE REVIEW

File: common/oidc_tokens.go [hunk 1]
```diff
@@ -26,8 +26,12 @@ import (
 )
 
 const (
+	ClaimClientIdOpenZiti = "openziti"
 	ClaimAudienceOpenZiti = "openziti"
 
+	//ClaimLegacyNative - to remove after SDKs stop using this as a client id
+	ClaimLegacyNative = "native"
```
Comment (dovholuknf) [inline common/oidc_tokens.go:33]: is there an issue to do this? as a reminder?

Comment (andrewpmartinez) [reply common/oidc_tokens.go:33]: An issue might be part of the solution. However, we lack a process for actually ensuring they are addressed.

File: controller/oidc_auth/storage.go [hunk 2]
```diff
@@ -894,15 +893,25 @@ func (s *HybridStorage) ValidateTokenExchangeRequest(_ context.Context, request
...
 	}
 
-	request.SetCurrentScopes(allowedScopes)
+	scopes := request.GetScopes()
+
+	if len(scopes) == 1 && scopes[0] == "" {
+		//no scopes supplied
+	} else {
+		for _, scope := range request.GetScopes() {
+			if scope != oidc.ScopeOfflineAccess && scope != oidc.ScopeOpenID {
```
Comment (dovholuknf) [inline controller/oidc_auth/storage.go:908]: seems like this if could be refactored mebbe?

File: router/state/manager.go [hunk 3]
```diff
@@ -475,8 +475,8 @@ func (sm *ManagerImpl) GetApiSession(token string) *ApiSession {
 		jwtToken, accessClaims, err := sm.ParseJwt(token)
 
 		if err == nil {
-			if !accessClaims.HasAudience(common.ClaimAudienceOpenZiti) {
-				pfxlog.Logger().Errorf("provided a token with invalid audience '%s', expected: %s", accessClaims.Audience, common.ClaimAudienceOpenZiti)
+			if !accessClaims.HasAudience(common.ClaimAudienceOpenZiti) || !accessClaims.HasAudience(common.ClaimLegacyNative) {
```
Comment (dovholuknf) [inline router/state/manager.go:478]: would it make sense at all to log when it's 'native' to let people know they have a problem? or will that sort of thing happen in another stage maybe?

Comment (dovholuknf) [reply router/state/manager.go:478]: also somewhat sad that the code isn't DRY - i take it they are three different structs?

File: tests/auth_oidc_refresh_test.go [hunk 4]
```diff
@@ -96,6 +172,31 @@ func Test_Authenticate_OIDC_Refresh(t *testing.T) {
...
+
+				newRefreshClaims := &jwt.MapClaims{}
+				newRefreshToken, _, err := jwtParser.ParseUnverified(newTokens.OidcTokens.RefreshToken, newRefreshClaims)
+				ctx.Req.NoError(err)
+				ctx.Req.NotNil(newRefreshToken)
+				newRefreshAudiences, err := newRefreshToken.Claims.GetAudience()
+				ctx.Req.NoError(err)
+				ctx.Req.Len(newRefreshAudiences, 1, "expected 1 audience")
+				ctx.Equal(originalRefreshAudiences[0], newRefreshAudiences[0])
+			})
```
Comment (dovholuknf) [inline tests/auth_oidc_refresh_test.go:199]: no test for ClaimLegacyNative in the tests. just making sure that's on purpose?

Comment (dovholuknf) [reply tests/auth_oidc_refresh_test.go:199]: i don't see any negative tests right?

File: common/oidc_tokens.go (hunk 1 above)
Comment (dovholuknf) [reply common/oidc_tokens.go:33]: sure. i was thinking to make it and link it here ? ¯\\_(ツ)_/¯

File: router/state/manager.go (hunk 3 above)
Comment (andrewpmartinez) [reply router/state/manager.go:478]: >DRY 
How would you change it? I don't see an issue.

> sense at all to log
I don't see why logging would do anything to help at this point in time. The goal is to remove the client id native. I am simply not breaking the test HA environment. The goal is to remove native once the C SDK is updated.

File: tests/auth_oidc_refresh_test.go (hunk 4 above)
Comment (andrewpmartinez) [reply tests/auth_oidc_refresh_test.go:199]: Because it is going to be removed here shortly.

File: controller/oidc_auth/storage.go (hunk 2 above)
Comment (andrewpmartinez) [reply controller/oidc_auth/storage.go:908]: Oh, the empty block. No that is an issuue.

File: common/oidc_tokens.go (hunk 1 above)
Comment (andrewpmartinez) [reply common/oidc_tokens.go:33]: Added https://github.com/openziti/ziti/issues/2307 to the pile.

File: router/state/manager.go (hunk 3 above)
Comment (andrewpmartinez) [reply router/state/manager.go:478]: After talking, these concerns are addressed when `native` support is removed.

**Review by dovholuknf — APPROVED** [review]

**Pull request merged**